### ISSUE TITLE
Typedef enforcement for variables that contain a function

### ISSUE THREAD

**Issue by StevenDoesStuffs** [palantir/tslint#2654]
### <s>Bug Report</s> Feature request

- __TSLint version__: 5.1.0
- __TypeScript version__: 2.3.2
- __Running TSLint via__: Webstorm (doesn't matter)

Feature (that is being requested): TSLint shouldn't ask for a type def for a variable that contains a function if the function is typed when variable-declaration in typedef is enabled. (With variable-declaration and the function types enforced, TSLint wants the function and the variable to be typed, and that is too repetitive for most)

#### TypeScript code being linted

```ts
let f = function(num: number): number {
    const i: number = num++;
    return i * i;
};
```

with `tslint.json` configuration: https://pastebin.com/wM7WDnvU (it's a bit long)

#### Actual behavior
Tells me I need a type declaration for f

#### <s>Expected</s> Requested behavior
Doesn't tell me that I need a typedef for f and allows it to be infered since it's already declared in the function (parameter and return types)

#### Rationale: 
There are going to be lots of people using tslint who want their variables and their functions to be typed, so most of them will enable typedef with the appropriate settings. Most of them probably don't want to typedef their functions twice like so: 
```ts
let f: (num: number) => number = function(num: number): number {
    const i: number = num++;
    return i * i;
};
```
Obviously, the parameter is duplicated (ish), the parameter type is dupliated, and the return type is also duplicated. In most cases, this is not the desirable and is considered overly verbose. Conclusion: it would be great if tslint can enforce typedefs for variables and functions but not in cases like the on above.

Note: I have a feeling this is already implemented and I'm just missing something, so I apologize in advance if that is the case.

Note2: Might be connected to #711

**Comment by MrAndersen1** [comment]
I agree with this. I want variables to be typedef'd but when you get to functions it becomes more cumbersome, less readable and also more error prone, even, when you begin renaming things.
```
const isMine = (a: Mine | Yours | Theirs): a is mine => a.type === "mine" ;
```
becomes
```
const isMine: (a: Mine | Yours | Theirs) => a is mine = (a: Mine | Yours | Theirs): a is mine => a.type === "mine" ;
```
Typing up a function certainly has its use-cases but when you never re-assign them then you'd be happy for an option to the rule so you don't feel compelled to type them up without ignore statements in your code.

**Comment by michaelw85** [comment]
+1 Completely agree. 

The readability for arrow functions decreases when typed on a single line. Having to define separate types to overcome this issue is quite annoying since it does not seem to add any value to the code quality in my opinion. 

I would like to enforce typedef for variables but not for arrow functions.

**Comment by gstamac** [comment]
Maybe it would be a good idea to have an option to enforce typedef for `let` and `var` definitions but not for `const`. You would probably use `const` in above example anyway. I would be willing to add a MR but need to discuss options first. I'm thinking about two different options `const-declaration` and `allow-const`. It would be probably a better idea to add `allow-const` so it doesn't break existing definition. It would be disabled by default.

**Comment by michaelw85** [comment]
Defining a rule for const seems like it would resolve this issue merely as a side effect. If you would introduce new rules why not one for this exact use case regardless if it's assigned to var/let/const

**Comment by gstamac** [comment]
@michaelw85 you're right but I don't see an easy solution. There is one more case - the constructor assignment - described in [711](https://github.com/palantir/tslint/issues/711#issuecomment-174917706). And I'm thinking of another ugly example
```
const x: {
prop1: string,
prop2: number,
prop3: boolean
} = {
prop1: 'str',
prop2: 3,
prop3: true
}

**Comment by michaelw85** [comment]
I guess this does not fall into the var rule but I would assume if you don't want typedef for arrow functions for vars you also don't want them when used in classes.

```ts
class MyClass { 
  public myArrowFunc: () => void = (): void => { 
  }
}
```
